libcontainer/nsenter/nsexec.c

```diff
@@ -2,351 +2,350 @@
-			      &ca);
+	int child;
+	struct clone_arg ca = {
+		.env = env,
+	};
+
+	child = clone(child_func, ca.stack_ptr, CLONE_PARENT | SIGCHLD | flags, &ca);
+
+	/*
+	 * On old kernels, CLONE_PARENT didn't work with CLONE_NEWPID, so we have
+	 * to unshare(2) before clone(2) in order to do this.
+	 */
+	if (errno == EINVAL) {
+		if (unshare(flags) < 0)
+			bail("unable to unshare namespaces");
+		child = clone(child_func, ca.stack_ptr, SIGCHLD | CLONE_PARENT, &ca);
 	}
+
 	return child;
 }
 
-// get init pipe from the parent. It's used to read bootstrap data, and to
-// write pid to after nsexec finishes setting up the environment.
-static int get_init_pipe()
+/*
+ * Gets the init pipe fd from the environment, which is used to read the
+ * bootstrap data and tell the parent what the new pid is after we finish
+ * setting up the environment.
+ */
+static int initpipe(void)
 {
-	char	buf[PATH_MAX];
-	char	*initpipe;
-	int	pipenum = -1;
+	int pipenum;
+	char *initpipe, *endptr;
 
 	initpipe = getenv("_LIBCONTAINER_INITPIPE");
-	if (initpipe == NULL) {
+	if (initpipe == NULL || *initpipe == '\0')
 		return -1;
-	}
 
-	pipenum = atoi(initpipe);
-	snprintf(buf, sizeof(buf), "%d", pipenum);
-	if (strcmp(initpipe, buf)) {
-		pr_perror("Unable to parse _LIBCONTAINER_INITPIPE");
-		exit(1);
-	}
+	pipenum = strtol(initpipe, &endptr, 10);
+	if (*endptr != '\0')
+		bail("unable to parse _LIBCONTAINER_INITPIPE");
 
 	return pipenum;
 }
 
-// num_namespaces returns the number of additional namespaces to setns. The
-// argument is a comma-separated string of namespace paths.
-static int num_namespaces(char *nspaths)
-{
-	int i;
-	int size = 0;
-
-	for (i = 0; nspaths[i]; i++) {
-		if (nspaths[i] == ',') {
-			size += 1;
-		}
-	}
-
-	return size + 1;
-}
-
 static uint32_t readint32(char *buf)
 {
-    return *(uint32_t *)buf;
+	return *(uint32_t *) buf;
 }
 
 static uint8_t readint8(char *buf)
 {
-    return *(uint8_t *)buf;
+	return *(uint8_t *) buf;
 }
 
-static void update_process_idmap(char *pathfmt, int pid, char *map, int map_len)
+static int write_file(char *data, size_t data_len, char *pathfmt, ...)
 {
-	char    buf[PATH_MAX];
-	int	len;
-	int	fd;
-
-	len = snprintf(buf, sizeof(buf), pathfmt, pid);
-	if (len < 0) {
-		pr_perror("failed to construct '%s' for %d", pathfmt, pid);
-		exit(1);
-	}
+	int fd, len;
+	char path[PATH_MAX];
+
+	va_list ap;
+	va_start(ap, pathfmt);
+	len = vsnprintf(path, PATH_MAX, pathfmt, ap);
+	va_end(ap);
+	if (len < 0)
+		return -1;
 
-	fd = open(buf, O_RDWR);
-	if (fd == -1) {
-		pr_perror("failed to open %s", buf);
-		exit(1);
-	}
+	fd = open(path, O_RDWR);
+	if (fd < 0)
+		return -1;
 
-	len = write(fd, map, map_len);
-	if (len == -1) {
-		pr_perror("failed to write to %s", buf);
-		close(fd);
-		exit(1);
-	} else if (len != map_len) {
-		pr_perror("Failed to write data to %s (%d/%d)",
-			buf, len, map_len);
-		close(fd);
-		exit(1);
-	}
+	len = write(fd, data, data_len);
+	if (len != data_len)
+		return -1;
```

Comment:
I'm currently thinking about that. In all of the error cases, we are going to exit so I don't think it _really_ matters that we close all of the file descriptors. I'm currently trying to figure out how `alloca` interacts with `clone` (hint: the answer is "not nicely") but I might also end up just adding a memory leak because we're going to `execve` anyway (which clears all of the allocated memory from the old process anyway).